File: desktop/internal/currency/currency.go
package currency

import (
	"fmt"
	"github.com/shopspring/decimal"
)

// Currency represents a monetary value with proper decimal precision
type Currency struct {
	value decimal.Decimal
}

// NewFromFloat creates a Currency from a float64 (use carefully!)
// This should only be used when reading from legacy DBF files
func NewFromFloat(f float64) Currency {
	// Round to 2 decimal places for currency
	d := decimal.NewFromFloat(f).Round(2)
	return Currency{value: d}
}

// NewFromString creates a Currency from a string representation
// This is the preferred way to create Currency values
func NewFromString(s string) (Currency, error) {
	d, err := decimal.NewFromString(s)
	if err != nil {
		return Currency{}, err
	}
	// Round to 2 decimal places for currency
	return Currency{value: d.Round(2)}, nil
}

// NewFromCents creates a Currency from an integer number of cents
func NewFromCents(cents int64) Currency {
	d := decimal.NewFromInt(cents).Div(decimal.NewFromInt(100))
	return Currency{value: d}
}

// Zero returns a zero Currency value
func Zero() Currency {
	return Currency{value: decimal.Zero}
}

// Add adds two Currency values
func (c Currency) Add(other Currency) Currency {
	return Currency{value: c.value.Add(other.value)}
}

// Sub subtracts a Currency value from another
func (c Currency) Sub(other Currency) Currency {
	return Currency{value: c.value.Sub(other.value)}
}

// Mul multiplies a Currency by a number
func (c Currency) Mul(multiplier decimal.Decimal) Currency {
	return Currency{value: c.value.Mul(multiplier).Round(2)}
}

// Div divides a Currency by a number
func (c Currency) Div(divisor decimal.Decimal) Currency {
	return Currency{value: c.value.Div(divisor).Round(2)}
}

// Neg returns the negative of a Currency value
func (c Currency) Neg() Currency {
	return Currency{value: c.value.Neg()}
}

// Abs returns the absolute value of a Currency
func (c Currency) Abs() Currency {
	return Currency{value: c.value.Abs()}
}

// IsPositive returns true if the Currency is positive
func (c Currency) IsPositive() bool {
	return c.value.IsPositive()
}

// IsNegative returns true if the Currency is negative
func (c Currency) IsNegative() bool {
	return c.value.IsNegative()
}

// IsZero returns true if the Currency is zero
func (c Currency) IsZero() bool {
	return c.value.IsZero()
}

// GreaterThan returns true if c > other
func (c Currency) GreaterThan(other Currency) bool {
	return c.value.GreaterThan(other.value)
}

// LessThan returns true if c < other
func (c Currency) LessThan(other Currency) bool {
	return c.value.LessThan(other.value)
}

// Equal returns true if c == other
func (c Currency) Equal(other Currency) bool {
	return c.value.Equal(other.value)
}

// ToCents returns the Currency value as integer cents
func (c Currency) ToCents() int64 {
	return c.value.Mul(decimal.NewFromInt(100)).IntPart()
}

// ToFloat64 returns the Currency value as a float64
// Use with caution - this can introduce precision errors
func (c Currency) ToFloat64() float64 {
	f, _ := c.value.Float64()
	return f
}

// ToString returns the Currency value as a string with 2 decimal places
func (c Currency) ToString() string {
	return c.value.StringFixed(2)
}

// ToJSON returns the Currency value suitable for JSON marshaling
// Returns a string to preserve precision
func (c Currency) ToJSON() string {
	return c.value.StringFixed(2)
}

// String implements the Stringer interface
func (c Currency) String() string {
	return fmt.Sprintf("$%s", c.value.StringFixed(2))
}

// ParseFromDBF parses a value from a DBF file into Currency
// Handles various types that DBF files might use for numeric values
func ParseFromDBF(value interface{}) Currency {
	if value == nil {
		return Zero()
	}
	
	switch v := value.(type) {
	case float64:
		return NewFromFloat(v)
	case float32:
		return NewFromFloat(float64(v))
	case int:
		return NewFromCents(int64(v * 100))
	case int64:
		return NewFromCents(v * 100)
	case string:
		// Try to parse as decimal
		if c, err := NewFromString(v); err == nil {
			return c
		}
		return Zero()
	default:
		return Zero()
	}
}

// SumCurrencies sums a slice of Currency values
func SumCurrencies(values []Currency) Currency {
	sum := Zero()
	for _, v := range values {
		sum = sum.Add(v)
	}
	return sum
}

File: desktop/frontend/src/utils/currency.ts
import Decimal from 'decimal.js';

// Configure Decimal for currency calculations (2 decimal places)
Decimal.set({ precision: 20, rounding: Decimal.ROUND_HALF_UP });

/**
 * Currency class for handling monetary values with proper decimal precision
 * Prevents floating-point arithmetic errors in financial calculations
 */
export class Currency {
  private value: Decimal;

  constructor(value: Decimal | number | string) {
    if (value instanceof Decimal) {
      this.value = value.toDecimalPlaces(2);
    } else {
      this.value = new Decimal(value).toDecimalPlaces(2);
    }
  }

  /**
   * Create a Currency from a number (use carefully with DBF data)
   */
  static fromNumber(num: number): Currency {
    return new Currency(num);
  }

  /**
   * Create a Currency from a string representation
   */
  static fromString(str: string): Currency {
    return new Currency(str);
  }

  /**
   * Create a Currency from cents
   */
  static fromCents(cents: number): Currency {
    return new Currency(new Decimal(cents).dividedBy(100));
  }

  /**
   * Create a zero Currency value
   */
  static zero(): Currency {
    return new Currency(0);
  }

  /**
   * Parse a value from DBF or API response
   */
  static parse(value: any): Currency {
    if (value === null || value === undefined || value === '') {
      return Currency.zero();
    }
    
    if (typeof value === 'number') {
      return Currency.fromNumber(value);
    }
    
    if (typeof value === 'string') {
      // Remove currency symbols and commas
      const cleaned = value.replace(/[$,]/g, '').trim();
      if (cleaned === '' || cleaned === '-') {
        return Currency.zero();
      }
      try {
        return Currency.fromString(cleaned);
      } catch {
        return Currency.zero();
      }
    }
    
    return Currency.zero();
  }

  /**
   * Add two Currency values
   */
  add(other: Currency): Currency {
    return new Currency(this.value.plus(other.value));
  }

  /**
   * Subtract a Currency value
   */
  subtract(other: Currency): Currency {
    return new Currency(this.value.minus(other.value));
  }

  /**
   * Multiply by a number
   */
  multiply(factor: number | Decimal): Currency {
    return new Currency(this.value.times(factor));
  }

  /**
   * Divide by a number
   */
  divide(divisor: number | Decimal): Currency {
    return new Currency(this.value.dividedBy(divisor));
  }

  /**
   * Get the negative of this Currency
   */
  negate(): Currency {
    return new Currency(this.value.negated());
  }

  /**
   * Get the absolute value
   */
  abs(): Currency {
    return new Currency(this.value.abs());
  }

  /**
   * Check if positive
   */
  isPositive(): boolean {
    return this.value.greaterThan(0);
  }

  /**
   * Check if negative
   */
  isNegative(): boolean {
    return this.value.lessThan(0);
  }

  /**
   * Check if zero
   */
  isZero(): boolean {
    return this.value.isZero();
  }

  /**
   * Compare with another Currency
   */
  greaterThan(other: Currency): boolean {
    return this.value.greaterThan(other.value);
  }

  /**
   * Compare with another Currency
   */
  lessThan(other: Currency): boolean {
    return this.value.lessThan(other.value);
  }

  /**
   * Check equality with another Currency
   */
  equals(other: Currency): boolean {
    return this.value.equals(other.value);
  }

  /**
   * Convert to cents (integer)
   */
  toCents(): number {
    return this.value.times(100).toNumber();
  }

  /**
   * Convert to number (use with caution)
   */
  toNumber(): number {
    return this.value.toNumber();
  }

  /**
   * Convert to string with 2 decimal places
   */
  toString(): string {
    return this.value.toFixed(2);
  }

  /**
   * Format as currency string with dollar sign
   */
  format(): string {
    const isNegative = this.isNegative();
    const absValue = this.abs().toString();
    
    // Add thousand separators
    const parts = absValue.split('.');
    parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    const formatted = parts.join('.');
    
    return isNegative ? `-$${formatted}` : `$${formatted}`;
  }

  /**
   * Format for display with optional color coding
   */
  formatWithColor(): { value: string; className: string } {
    const formatted = this.format();
    const className = this.isNegative() ? 'text-red-600' : 'text-green-600';
    return { value: formatted, className };
  }

  /**
   * Create a Currency from a JSON value
   */
  static fromJSON(json: any): Currency {
    return Currency.parse(json);
  }

  /**
   * Convert to JSON-serializable value
   */
  toJSON(): string {
    return this.toString();
  }
}

/**
 * Sum an array of Currency values
 */
export function sumCurrencies(values: Currency[]): Currency {
  return values.reduce(
    (sum, value) => sum.add(value),
    Currency.zero()
  );
}

/**
 * Calculate the average of Currency values
 */
export function averageCurrencies(values: Currency[]): Currency {
  if (values.length === 0) {
    return Currency.zero();
  }
  const sum = sumCurrencies(values);
  return sum.divide(values.length);
}

/**
 * Find the maximum Currency value
 */
export function maxCurrency(values: Currency[]): Currency {
  if (values.length === 0) {
    return Currency.zero();
  }
  return values.reduce((max, value) => 
    value.greaterThan(max) ? value : max
  );
}

/**
 * Find the minimum Currency value
 */
export function minCurrency(values: Currency[]): Currency {
  if (values.length === 0) {
    return Currency.zero();
  }
  return values.reduce((min, value) => 
    value.lessThan(min) ? value : min
  );
}

/**
 * Format a number as currency (legacy support)
 * @deprecated Use Currency class instead
 */
export function formatCurrency(value: number | string | null | undefined): string {
  return Currency.parse(value).format();
}

/**
 * Parse and calculate a GL balance using proper decimal arithmetic
 */
export function calculateGLBalance(
  debits: number | string,
  credits: number | string,
  accountType: number
): Currency {
  const totalDebits = Currency.parse(debits);
  const totalCredits = Currency.parse(credits);
  
  // Account Types:
  // 1 = Assets (Debit normal balance)
  // 2 = Liabilities (Credit normal balance)
  // 3 = Equity (Credit normal balance)
  // 4 = Revenue/Income (Credit normal balance)
  // 5 = Expenses (Debit normal balance)
  
  switch (accountType) {
    case 1: // Assets
    case 5: // Expenses
      return totalDebits.subtract(totalCredits);
    case 2: // Liabilities
    case 3: // Equity
    case 4: // Revenue
      return totalCredits.subtract(totalDebits);
    default:
      // Default to asset behavior
      return totalDebits.subtract(totalCredits);
  }
}